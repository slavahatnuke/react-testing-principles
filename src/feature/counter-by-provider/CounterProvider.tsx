import React, {createContext, ReactNode, useContext, useState} from "react";

export type ICounterContext = {
  counter: number,
  increment: () => void,
  decrement: () => void
};

export const CounterContext = createContext<ICounterContext>(null as any)

export const CounterProvider = ({children, value}: { children: ReactNode, value?: Partial<ICounterContext> }) => {
  const [counter, setCounter] = useState(value?.counter ?? 0);

  const increment = () => {
    setCounter((prevCounter) => prevCounter + 1);
  };

  const decrement = () => {
    setCounter((prevCounter) => prevCounter - 1);
  };

  const context: ICounterContext = {
    counter,
    increment,
    decrement
  }

  return <CounterContext.Provider value={context}>{children}</CounterContext.Provider>;
}

export const useCounterByProvider = () => {
  return useContext(CounterContext);
}