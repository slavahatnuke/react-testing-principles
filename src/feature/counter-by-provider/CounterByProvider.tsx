import {CounterView} from "../counter/CounterView";
import {useCounterByProvider} from "./CounterProvider";

type ICounterProps = {};

// Composition of logic and view
export const CounterByProvider = ({}: ICounterProps) => {
  // counter logic
  const {counter, increment, decrement} = useCounterByProvider();

  // counter view
  return <CounterView counter={counter} increment={increment} decrement={decrement}/>;
};


