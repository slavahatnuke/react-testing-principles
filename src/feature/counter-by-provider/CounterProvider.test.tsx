import {act, renderHook} from '@testing-library/react-hooks'
import {CounterProvider, useCounterByProvider} from "./CounterProvider";
import renderer from "react-test-renderer";
import React from "react";
import {CounterByProvider} from "./CounterByProvider";

// test examples and the following links:

// https://github.com/testing-library/react-hooks-testing-library
// https://testing-library.com/docs/example-react-context/
// https://react-hooks-testing-library.com/usage/advanced-hooks#context

describe(CounterProvider.name, () => {
  test(useCounterByProvider.name, async () => {

    const {result} = renderHook(() => useCounterByProvider(), {
      wrapper: CounterProvider,
    })

    expect(result.current.counter).toBe(0)

    act(() => {
      result.current.increment()
    })

    expect(result.current.counter).toBe(1)

    act(() => {
      result.current.decrement()
    })

    expect(result.current.counter).toBe(0)
  })

  test('snapshot by provider', () => {
    const component1 = renderer.create(
      <CounterProvider>
        <CounterByProvider/>
      </CounterProvider>,
    );

    expect(component1.toJSON()).toMatchSnapshot();

    const component2 = renderer.create(
      <CounterProvider value={{counter: 10}}>
        <CounterByProvider/>
      </CounterProvider>,
    );
    expect(component2.toJSON()).toMatchSnapshot();
  })
})

