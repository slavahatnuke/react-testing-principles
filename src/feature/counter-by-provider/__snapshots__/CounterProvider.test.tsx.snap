// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterProvider snapshot by provider 1`] = `
Array [
  <h3
    data-testid="counter-header"
  >
    Counter (
    0
    )
  </h3>,
  <button
    data-testid="increment"
    onClick={[Function]}
  >
    +
  </button>,
  <button
    data-testid="decrement"
    onClick={[Function]}
  >
    -
  </button>,
]
`;

exports[`CounterProvider snapshot by provider 2`] = `
Array [
  <h3
    data-testid="counter-header"
  >
    Counter (
    10
    )
  </h3>,
  <button
    data-testid="increment"
    onClick={[Function]}
  >
    +
  </button>,
  <button
    data-testid="decrement"
    onClick={[Function]}
  >
    -
  </button>,
]
`;
